pkg/skaffold/kubernetes/logger/log.go

```diff
@@ -77,61 +92,65 @@ func (a *LogAggregator) SetSince(t time.Time) {
 	a.sinceTime = t
 }
 
-// Start starts a logger that listens to pods and tail their logs
+// StartLogger starts a logger that listens to pods and tail their logs
 // if they are matched by the `podSelector`.
-func (a *LogAggregator) Start(ctx context.Context, namespaces []string) error {
+func (a *LogAggregator) StartLogger(ctx context.Context, out io.Writer, namespaces []string) error {
+	var err error
 	if a == nil {
 		// Logs are not activated.
 		return nil
 	}
+	a.startOnce.Do(func() {
+		a.output = out
+
+		a.podWatcher.Register(a.events)
+		stopWatcher, err := a.podWatcher.Start(namespaces)
+		a.stopWatcher = stopWatcher
+		if err != nil {
+			return
+		}
 
-	a.podWatcher.Register(a.events)
-	stopWatcher, err := a.podWatcher.Start(namespaces)
-	if err != nil {
-		return err
-	}
-
-	go func() {
-		defer stopWatcher()
-
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			case evt, ok := <-a.events:
-				if !ok {
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
 					return
-				}
+				case evt, ok := <-a.events:
+					if !ok {
+						return
+					}
 
-				// TODO(dgageot): Add EphemeralContainerStatuses
-				pod := evt.Pod
-				for _, c := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
-					if c.ContainerID == "" {
-						if c.State.Waiting != nil && c.State.Waiting.Message != "" {
-							output.Red.Fprintln(a.output, c.State.Waiting.Message)
+					// TODO(dgageot): Add EphemeralContainerStatuses
+					pod := evt.Pod
+					for _, c := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
+						if c.ContainerID == "" {
+							if c.State.Waiting != nil && c.State.Waiting.Message != "" {
+								output.Red.Fprintln(a.output, c.State.Waiting.Message)
+							}
+							continue
 						}
-						continue
-					}
 
-					if !a.trackedContainers.add(c.ContainerID) {
-						go a.streamContainerLogs(ctx, pod, c)
+						if !a.trackedContainers.add(c.ContainerID) {
+							go a.streamContainerLogs(ctx, pod, c)
+						}
 					}
 				}
 			}
-		}
-	}()
-
-	return nil
+		}()
+	})
+	return err
 }
 
-// Stop stops the logger.
-func (a *LogAggregator) Stop() {
+// StopLogger stops the logger.
+func (a *LogAggregator) StopLogger() {
 	if a == nil {
 		// Logs are not activated.
 		return
 	}
-
-	close(a.events)
+	a.stopOnce.Do(func() {
```

Comment:
wdyt about keeping a count of the `Start` invocations and closing the channel only when the corresponding number of `Stop` have been called?